charms/istio-pilot/src/charm.py

```diff
@@ -828,7 +894,97 @@ def _istiod_svc(self):
+            return {
+                "ssl-crt": base64.b64encode(
+                    ssl_secret.get_content()["ssl-crt"].encode("ascii")
+                ).decode("utf-8"),
+                "ssl-key": base64.b64encode(
+                    ssl_secret.get_content()["ssl-key"].encode("ascii")
+                ).decode("utf-8"),
+            }
+        except SecretNotFoundError:
+            return {
+                "ssl-crt": base64.b64encode(self._cert_handler.cert.encode("ascii")).decode(
+                    "utf-8"
+                ),
+                "ssl-key": base64.b64encode(self._cert_handler.key.encode("ascii")).decode(
+                    "utf-8"
+                ),
+            }
+
+    # ---- Start of the block
+    # ---- WARNING: this feature is not recommended, but is supported in 1.17 and 1.18.
+    # ---- For details please refer to canonical/istio-operators#380.
+    # ---- FIXME: Remove this block after releasing 1.18.
+    def _use_https(self) -> bool:
+        """Return True if only one of the TLS configurations are enabled, False if none are.
+
+        Raises:
+            ErrorWithStatus: if both configurations are enabled.
+        """
+        if self._use_https_with_tls_provider() and self._use_https_with_tls_secret():
+            self.log.info(
+                "Only one TLS configuration is supported at a time."
+                "Either remove the TLS certificate provider relation,"
+                "or the TLS manual configuration with the remove-tls-secret action."
+            )
+            raise ErrorWithStatus(
+                "Only one TLS configuration is supported at a time. See logs for details.",
+                BlockedStatus,
+            )
+        return self._use_https_with_tls_provider() or self._use_https_with_tls_secret()
+
+    def _use_https_with_tls_secret(self) -> bool:
+        """Return True if both SSL key and crt are set with save-tls-secret, False otherwise.
+
+        Raises:
+            ErrorWithStatus: if one of the values is missing.
+        """
+
+        # Ensure the secret that holds the values exist otherwise we can assume
+        # that users don't want to configure TLS or they want to do it via a TLS provider.
+        try:
+            secret = self.model.get_secret(label="istio-tls-secret")
+        except SecretNotFoundError:
+            return False
+
+        # Fail if ssl is only partly configured as this is probably a mistake
+        ssl_key = secret.get_content()["ssl-key"]
+        ssl_crt = secret.get_content()["ssl-crt"]
+        if _xor(ssl_key, ssl_crt):
+            missing = "ssl-key"
+            if not secret.get_content()["ssl-crt"]:
+                missing = "ssl-crt"
+            raise ErrorWithStatus(f"Missing {missing}, cannot configure TLS.", BlockedStatus)
+
+        # If both the SSL key and cert are provided, we configure TLS
+        return True if ssl_key and ssl_crt else False
```

Comment:
is this necessary now that you've added the `required` bit?

Reply:
Yes, the action won't do anything anyway. Thanks for the suggestion!